ENH: Optimize `np.empty` for scalar arguments

fandreuz opened the pull request:
This PR includes the modifications required to fix #19010, plus several additions discussed in the issue, like the usage of `PySequence_Fast` and the corresponding family of functions.

fandreuz committed:
ENH: New function 'PyArray_IntpFromScalar'. Detect scalars early on in 'PyArray_IntpConverter'

fandreuz committed:
ENH: Refactor code to use PySequence_Fast

seberg reviewed (commented):
The approach looks good, I made a few comments about smaller refactoring and as matti said, the test coverage must be improved for the error paths.

fandreuz committed:
use IntpFromSequence also in IntpFromIndexSequence loop

fandreuz committed:
better name for nd in scalar branch

fandreuz committed:
refactor PyArray_PyIntAsIntp -> intp_from_scalar: return actual value. move up. remove from header file

fandreuz committed:
refactor error handling

fandreuz committed:
collapse error handling in PySequence_Fast

fandreuz committed:
remove unused variable

fandreuz committed:
doxygen for PyArray_IntpFromSequence, convert to PySequence_Fast in PyArray_IntpFromSequence

fandreuz committed:
ENH: improve error handling

fandreuz committed:
TST: test PyArray_IntpConverter and PyArray_IntpFromSequence

fAndreuzzi commented:
Thank you @seberg, I added two minimal tests and it looks like the coverage is increased. Let me know if there's something missing.

I noticed that there are still several untested lines:
1. Is there a way to simulate a full memory in order to get them covered?
2. How should I "test" `Py_DECREF`?

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -125,27 +196,34 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
     if (len < 0) {
         PyErr_SetString(PyExc_TypeError,
                 "expected sequence object with len >= 0 or a single integer");
+        Py_DECREF(seq_obj);
```
`PySequence_Fast_GET_SIZE` will never fail. You can remove the whole `< 0` and `== -1` blocks.  Any failure will have occurred in `PySequence_Fast` already.

We may have to think briefly whether we have to fall back to the above branch if  `PySequence_Fast(...)` fails, so that it is a bit like a: `try: list(obj); except: int(obj)` path in Python.

seberg replied:
The reason why this line is not covered, is that is is impossible to cover it :).

fAndreuzzi replied:
I removed the redundant checks in e5b77c1.

> We may have to think briefly whether we have to fall back to the above branch if PySequence_Fast(...) fails

Which are the conditions under which `PySequence_Fast` might fail in this case?

> The reason why this line is not covered, is that is is impossible to cover it :).

Got it :)

seberg replied:
`PySequence_Fast` should fail exactly in the same cases as `list(obj)` failing.

fAndreuzzi replied:
Isn't this check sufficient to ensure that `PySequence_Fast` does not fail?
https://github.com/numpy/numpy/blob/5cbde7fef71268fc0b834c5f82aee798c6cb0792/numpy/core/src/multiarray/conversion_utils.c#L152

seberg replied:
No, for one, `PySequence_Check` only checks very quickly if it should be a squence.  But sequences could be badly implemented and only error when you get the elements.

The one problem that might occur is to try the sequence conversion first. Then we will still try the `int(seq)` path for broken sequences.  This is probably actually important!  Could you add a test for `np.array(3)` being accepted?  
```
seq = NULL
if not Long_CheckExact and PySequence_Check:
   seq = ...
   if seq == NULL:
       PyErr_Clear()  # ignore the error.

if seq == NULL:
    # integer path

else:
   # sequence path
```
A bit annoying dance, but I think it ensures we don't accidentally change more behaviour than anticipated. The `PyErr_Clear()` makes it a `try/except` clause.

fAndreuzzi replied:
Please @seberg have a look at 6da32a8. 

For some reason now the following test fails (everything else is fine):
```
    def test_none(self):
        # once the warning expires, this will raise TypeError
        with pytest.warns(DeprecationWarning):
>           assert self.conv(None) == ()
E           assert None == ()
E            +  where None = <built-in function run_intp_converter>(None)
E            +    where <built-in function run_intp_converter> = <numpy.core.tests.test_conversion_utils.TestIntpConverter object at 0x7fda6307ecd0>.conv

self       = <numpy.core.tests.test_conversion_utils.TestIntpConverter object at 0x7fda6307ecd0>

numpy/core/tests/test_conversion_utils.py:190: AssertionError
```

From what I understand, this case should be handled in:
https://github.com/numpy/numpy/blob/6da32a8349eba9f1dde56c12bdc347eda4ded5c7/numpy/core/src/multiarray/conversion_utils.c#L120-L128
But it looks like something went wrong.

fAndreuzzi replied:
@seberg The failing test has been fixed in 83d8d59.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,7 +145,43 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+    PyObject *seq_obj = PySequence_Fast(obj,
+        "An error occurred during a call to `PySequence_Fast`, an iterable "
+        "object is required. The given object is not a scalar integer nor a "
+        "sequence.");
```
This error message (and also below) is not user-friendly. Maybe check what it is, and what it was.  If we end up clearing it (due to a try/except), you can leave it empty I think.  Otherwise, it should be something like `expected an integer or a sequence of integers`.

fAndreuzzi replied:
Fixed in 12c9b4e

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -1028,7 +1094,28 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
+    PyObject *seq_obj = PySequence_Fast(seq,
+        "An error occurred during a call to `PySequence_Fast`, an iterable "
+        "object is required. The given object is not a scalar integer nor a "
+        "sequence.");
```
Same here, this is fine, but the error message is not helpful.  Maybe the default is even OK (you could compare what it was before and after for an invalid object?).

fAndreuzzi replied:
Fixed in 12c9b4e.

I was not able to recover the default error message since `PySequence_Fast()` strongly requires the message, which means that passing `NULL` results in a segmentation fault, and passing `""` results in `numpy/core/tests/test_multiarray.py:526: TypeError`, which does not contain any error message.

I replaced my message with yours (slightly modified).

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,7 +145,43 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+        }
+        seq->len = len;
+
+        npy_intp rvalue = intp_from_scalar(obj, (npy_intp *)seq->ptr, 0);
```
Could you rewrite this so that it reads:
```
seq->ptr[0] = intp_from_scalar(obj);
if (error_converting(seq->ptr[0]) {
    npy_free_cache_dim_obj(*seq);
    seq->ptr = NULL;  /* I would think this is unnecessary, but keep it if it was there? */
    return NPY_FAIL;
}
```

seberg replied:
I.e. I am thinking that you should simplify `intp_from_scalar`.

fAndreuzzi replied:
I thought for some minutes about the fact that `intp_from_scalar` was too verbose (to call, and also in its implementation), but then I decided to stick with the "format" of the other functions.

Fixed in 8dc0f00

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,7 +145,43 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+        "object is required. The given object is not a scalar integer nor a "
+        "sequence.");
+    if (seq_obj == NULL) {
+        return NPY_FAIL;
```
Missing the cleanup:
```
            npy_free_cache_dim_obj(*seq);
```
(and possibly setting `ptr` to NULL, but I don't think it should be necessary).

fAndreuzzi replied:
Fixed in 5ae9db6

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -125,27 +196,34 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
-        PyErr_Format(PyExc_ValueError, "maximum supported dimension for an ndarray is %d"
-                     ", found %d", NPY_MAXDIMS, len);
+        PyErr_Format(PyExc_ValueError, "maximum supported dimension for an "
+                    "ndarray is %d, found %d", NPY_MAXDIMS, len);
```
NIT: I/we currently prefer formatting these as (the continuation is 8 space indented)
```
PyErr_Format(PyExc_ValueError,
        "maximum supported dimension for an ndarray is %d, found %d",
        NPY_MAXDIMS, len);
```

fAndreuzzi replied:
Fixed in d549b0f, let me know if you spot something similar which I did not see.

seberg replied:
Oh, I was going for a line break after the `PyExc_ValueError`, keeps the string together.  To be fair, we only use that style fairly consistently for errors.

seberg reviewed (commented):
Thanks! Seems code coverage is great now.  I made some comments, I think the helper function would be better to work on a signle value.  I do have to look a bit closer at the logic (whether `PySequence_Check` is absolutely safe, or we should go back to a try/except style).

If you get stuck/fed up iterating, let me know and I can do some fixups for you.

fandreuz committed:
PySequence_Fast_GET_SIZE never fails. code/comments cleanup

fandreuz committed:
fix missing cleanup

fandreuz committed:
fix number of spaces

fandreuz committed:
simplify intp_from_scalar

fandreuz committed:
more user-friendly error message

fAndreuzzi commented:
Thank you @seberg! This is actually my first time using Python and C together (also, my experience with C is negligible), therefore your reviews are very useful.

fandreuz committed:
invert order of evaluation for sequences and scalars

fandreuz committed:
fix failing test

fandreuz committed:
test that np.array(3) is accepted

fAndreuzzi commented:
Hi @seberg, just wanted to notify you that I think I pushed all the changes you required yesterday (addition of a test for `np.array(3)` and the new sequence of evaluation (list, scalar, ...)

Let me know what you think and if there is something more I can do :)

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -1028,7 +1060,22 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
+            "Expected an integer or an iterable/sequence of integers.");
+    if (seq_obj == NULL) {
+        return -1;
+    }
```
Should we invert this here as well?  To make it "try/except style"?

fAndreuzzi replied:
I wasn't sure since in `PyArray_IntpFromSequence` we do not need to reserve memory "manually". I'm going to refactor this as well

fAndreuzzi replied:
Fixed in 952871653a474c61b81a286dc6797da75ca226e7, I also introduced two tests with an overflow error for more extensive checking.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -1028,7 +1060,22 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
+        vals[0] = intp_from_scalar(seq);
+        if (PyErr_Occurred() != NULL) {
+            return -1;
+        } else {
```
```suggestion
        }
        else {
```
Also in some other places.  It is fine, but just not NumPy C-style.

fAndreuzzi replied:
Fixed in 6180fb68b20a967163eadfa47a80c1b526a0afe2

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -1028,7 +1060,22 @@ PyArray_IntpFromIndexSequence(PyObject *seq, npy_intp *vals, npy_intp maxvals)
-    return PyArray_IntpFromIndexSequence(seq, vals, (npy_intp)maxvals);
+    if (PyLong_CheckExact(seq) || !PySequence_Check(seq)) {
+        vals[0] = intp_from_scalar(seq);
+        if (PyErr_Occurred() != NULL) {
```
```suggestion
        if (error_converting(vals[0])) {
```
Python (and we) ensure the result is -1 if an error occurred, `error_converting` uses that for the common-case.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -78,6 +78,29 @@ PyArray_OutputConverter(PyObject *object, PyArrayObject **address)
+        if (PyErr_ExceptionMatches(PyExc_OverflowError))
+            PyErr_SetString(PyExc_ValueError,
+                    "Maximum allowed dimension exceeded");
+        else
```
```suggestion
        }
        else {
```
and also other places.  NumPy C-style _never_ leaves out curly brackets.

fAndreuzzi replied:
Fixed in 6180fb68b20a967163eadfa47a80c1b526a0afe2

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -78,6 +78,29 @@ PyArray_OutputConverter(PyObject *object, PyArrayObject **address)
+intp_from_scalar(PyObject *ob)
+{
+    /*
+    * Convert the given value to an integer. Then store it and check for
+    * errors.
+    */
```
```suggestion
     * Convert the given value to an integer. Then store it and check for
     * errors.
     */
```
style nit: keep the asterisks/stars aligned.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -78,6 +78,29 @@ PyArray_OutputConverter(PyObject *object, PyArrayObject **address)
 }
 
+static npy_intp
+intp_from_scalar(PyObject *ob)
```
After thinking about, it maybe we can call it `dimension_from_scalar`, makes it a bit clearer where it is used (normally?) for, and why the error is: Maximum allowed dimension exceeded.

fAndreuzzi replied:
Fixed in fd1c1ceb6924cce4122bee14fe3b20f2f3ec2d8a

seberg commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -2363,6 +2363,35 @@ run_intp_converter(PyObject* NPY_UNUSED(self), PyObject *args)
+test_scalar_intp_from_sequence(PyObject *dummy, PyObject *obj)
+{
+    npy_intp vals[1];
+    if (vals == NULL) {
+        PyErr_NoMemory();
+        return NPY_FAIL;
+    }
```
```suggestion
```
This can't fail.  If you write something like `npy_intp vals[1]` C places it on the stack and not the heap.  That means there is no call to `malloc` that could fail.  (The "downside" is that the stack has a limited size, so you can't put can only put smallish amounts of data on it.)

seberg commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -2363,6 +2363,35 @@ run_intp_converter(PyObject* NPY_UNUSED(self), PyObject *args)
+test_scalar_intp_converter(PyObject *dummy, PyObject *obj)
+{
+    PyArray_Dims dummy_dims;
+    int output = PyArray_IntpConverter(obj, &dummy_dims);
```
For converters here, you could also write:
```
if (!PyArray_IntpConverter(obj, &dummy_dims)) {
    return NULL;
}
/* continue here (no else if there is an error return) */
```

fAndreuzzi replied:
I probably did not understand your review, I refactored the code to this form:
```c
PyObject *
test_scalar_intp_converter(PyObject *dummy, PyObject *obj)
{
    PyArray_Dims dummy_dims;
    if (!PyArray_IntpConverter(obj, &dummy_dims)) {
        return NULL;
    }
    else {
        return PyLong_FromLong(dummy_dims.ptr[0]);
    }
}
```

What do you mean by _no else if there is an error return_?

seberg replied:
This is fine, but you do not need the `else` clause at all, since the `if` already returns.  Preferences vary, although for short error-returns I think it is firmly on the side of removing the `else`.

seberg commented on numpy/core/src/multiarray/_multiarray_tests.c.src:
```diff
@@ -2363,6 +2363,35 @@ run_intp_converter(PyObject* NPY_UNUSED(self), PyObject *args)
+    if (output != NPY_SUCCEED) {
+        return NULL;
+    } else {
+        return PyLong_FromLong(dummy_dims.ptr[0]);
```
`dummy_dims.ptr` needs to be freed here!  So you must call `npy_free_cache_dim` (or just free).

In case you would want to test more (I am not sure it is really necessary for `PyArray_IntpConverter`), you could also use: `PyObject *result = PyArray_IntTupleFromIntp(dummy_dims.nd, dummy_dims.ptr)` then you got all of the result, and not just the first number.

fAndreuzzi replied:
Fixed in 3e6b91c5fd6f32364413c0a8eb783e92707e4328

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+        * Check whether the value set by intp_from_scalar is different than
+        * the value returned, which means that an error occurred.
+        */
+        if (PyErr_Occurred() != NULL) {
```
Same here, use `error_converting` largely for consistency.

fAndreuzzi replied:
Fixed in 5c5792e387c54e7fe4e3265bc8be49e3e5b97d47

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+        */
+        if (PyErr_Occurred() != NULL) {
+            npy_free_cache_dim_obj(*seq);
+            seq->ptr = NULL;
```
Above we have this error:
```
        PyErr_SetString(PyExc_TypeError,
                 "expected sequence object with len >= 0 or a single integer");
```
which is the "all else failed" path.  I would rmeove the `len >= 0` it doesn't make much sense.  But otherwise, it seems sensible to preserve it.  Could make it `expected a sequence of integers or a single integer`.

To be clear: This is basically the "all else failed" path here, since we now already:
* Checked whether it is a sequence (it may not contain integers, but we do not reach if it is a sequence).
* Checked whether it is an integer.

fAndreuzzi replied:
Fixed in c545c3dbca2f0d460d7a593c42654d4779c95dd3, is this what you expected?

fAndreuzzi replied:
Had to fix this a little bit in 9add7a7a025c454d68db9928da2e493137d959d9 since one of the tests which checks the returned exception when an overflow occurs expected a `ValueError` instead of a `TypeError`.

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+        /*
+        * obj has been recognized as an iterable of integer values
+        */
+
```
```suggestion
```
(and * indentation above)

fAndreuzzi replied:
Fixed in e8bd7532fe0c63d89d81a502b2c1c6ecc90b737e

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -110,42 +130,77 @@ PyArray_IntpConverter(PyObject *obj, PyArray_Dims *seq)
+               "Expected an integer or an iterable/sequence of integers.");
+        if (seq_obj == NULL) {
+            // we ignore this error since it is managed below
+            PyErr_Clear();
```
Ah, we should be able to cover this, but it is a bit tricky, something like:
```
class BadSequenceButInt:
    def __len__(self):
        return 4

    def __getitem__(self, item):
        raise ValueError("Bad sequence!")

    def __index__(self):
         return 2
```
I guess we change the behaviour of such a ridiculous object (seems it was previously rejected).  But I think that is good (or fine).

fAndreuzzi replied:
Should I ignore this one?

seberg replied:
Yes, fine to ignore it, I will add that `np.array(0)` test that I mentioned.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -449,6 +449,10 @@ def test_array(self):
 
+    def test_0d_array(self):
+        assert np.array(3) == 3
+        assert np.array(10).shape == ()
```
Ah, my "BadSequence" test is probably not necessary, then, I guess.  This is not what I was going for ;):
```
np.ones(np.array(3)).shape == (3,)
```
is what I was going for.  The reason is that `list(np.array(3))` fails, I expect `PySequence_Check(np.array(3))` is True.

seberg replied:
Testing `np.array(3)` in `test_scalar_intp_converter` and `test_scalar_intp_from_sequence ` would be perfect (and more than sufficient).

fAndreuzzi replied:
Tentative fix in 5d11ffc2614707fedd376ce668ec45c7cdb28880, what do you think?

My doubt is that testing `np.array(3)` is not suitable to be inserted inside `test_scalar_intp_converter ` or `test_scalar_intp_from_sequence` since only one path between the two possible paths (`PyArray_IntpFromSequence` or `PyArray_IntpConverter`) is followed, the other test is redundant (and wrongly located in some sense). Am I missing something?

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -511,6 +515,23 @@ def test_array_as_keyword(self, func):
+        assert test_scalar_intp_from_sequence(10) == 10
+        assert test_scalar_intp_from_sequence(-1) == -1
+
+        assert_raises(TypeError, test_scalar_intp_from_sequence, 'foo')
```
Both of these tests look exactly the same, so you could move the import to the stop, and then rewrite them as:
```
@pytest.mark.parametrize("converter", [test_scalar_intp_converter, test_scalar_intp_from_sequence])
```
But I guess if you end up changing/extending, it is just as well to keep it as two.

fAndreuzzi replied:
Sure, I'm just waiting to see the final form of `test_scalar_intp_converter` and `test_scalar_intp_from_sequence`

seberg commented on numpy/core/src/multiarray/conversion_utils.c:
```diff
@@ -956,65 +1011,42 @@ PyArray_PyIntAsIntp(PyObject *o)
-            }
+    Py_ssize_t nd = PySequence_Fast_GET_SIZE(seq);
+    PyObject *op;
+    for (Py_ssize_t i = 0; i < PyArray_MIN(nd,maxvals); i++) {
```
```suggestion
    for (Py_ssize_t i = 0; i < PyArray_MIN(nd, maxvals); i++) {
```
super nitpicky :)

seberg reviewed (commented):
Thanks, this is coming along nicely.  A few style nits that apply to some unmarked places also.  The test is not quite what I had in mind.  Otherwise, my only concern is that I think that public-API function also should get the "try/except" logic.

In any case, I love how much simpler `PyArray_IntpFromIndexSequence` is now.

fandreuz committed:
enforce NumPy C-style (curly braces, else on newline)

fandreuz committed:
refactor error check to uses error_converting

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

fandreuz committed:
align stars in comments

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

fandreuz committed:
intp_from_scalar -> dimension_from_scalar

fandreuz committed:
remove unused error check (memory allocation on stack)

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

fandreuz committed:
free dummy_dims.ptr. test_scalar_intp_converter: return value from PyArray_IntTupleFromIntp

fandreuz committed:
refactor error check

fandreuz committed:
fix star alignment

fandreuz committed:
fix spacing

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

fandreuz committed:
fix test cases (improve test_0d_array_shape)

fandreuz committed:
add error message to all-else-failed branch in PyArray_IntpConverter. improve error message in PyArray_IntpFromSequence

fandreuz committed:
fix error handling to deliver the proper exception when an overflow occurs

fandreuz committed:
refactor structure (try .. except) of PyArray_IntpFromSequence to match the PyArray_IntpConverter one. overflow tests

fAndreuzzi commented:
Hi @seberg, thank you again for your review, I think I addressed all your suggestions (with several questions).

fandreuz committed:
enforce else convention

seberg commented:
Depending on your system, a few things you could try to debug is:
* use a python debug version, e.g. debian has `python3-dbg` that you can just install and run as normal
* If you do not have a debug version, you can still try `PYTHONMALLOC=debug ...`, which is likely the relevant part that a debug build would give you.
* Run in `gdb` (but that is not super likely to help if you cannot reproduce the issue).
* Run in `valgrind` (usually simple, but very slow).  You have to use `PYTHONMALLOC=malloc valgrind ...`. Make sure to only run the specific test you are worried about.

If everything fails, we can run the (non windows) jobs using GDB in CI also.

fAndreuzzi commented:
Thank you @seberg, I used `gdb` several times to find bugs which I introduced during the last days, but unfortunately I cannot reproduce this bug on my computer. I can only test for its appearance by running GH via this PR.

I managed to discover that the problem disappears if I comment `free(dummy_dims.ptr);` in `test_scalar_intp_converter`.

seberg commented:
Sorry, my bad.  I keep forgetting that `PyDimMem_FREE` only ends up at `free` in non-debug mode. I will fix this and do some smaller cleanup's in a few minutes if you don't mind.

fAndreuzzi commented:
Sure! I cleared away my experiments from the branch

seberg committed:
MAINT: Small tweaks and fixups

Largely just my personal style preferences.  Other small changes:

* `PyErr_Clear()` is not necessary before a `PyErr_Set...`
* Rename testing functions, testing function starting with
  `test_` and imported top-level would get executed, just a possible
  cause of confusion.
* Removed/shortened/rephrased a few comments, nothing particular
  and probably very subjective.
* Use `PyLong_FromLongLong` instead of `FromLong`. The former is
  technically incorrect on windows (and may cause compiler warnings)
  This should be always correct...

seberg committed:
ENH,TST: Refactor error messages and tests

This modifies the error messages to give a more clear error
when an invalid object is passed in (neither a sequence of ints
nor an integer).  The message is now:

  expected a sequence of integers or a single integer, got '<repr>'

Moves tests, expand a little, and use parametrization.

seberg commented:
Also refactored the tests a bit and tweaked the error message.  Someone else may want to have a look, but I think it should be good to go.  The most common error message here is now:
```python
np.ones(object())
```
```python-traceback
TypeError: expected a sequence of integers or a single integer, got '<object object at 0x7f90d733c770>'
```
This could also make sense with exeption chaining, but I am not sure it is worth the trouble.  The error type should never change, only the message itself.

The code here effectively uses `try: tuple(obj): ... except: operator.index(obj)`, this is slightly more relaxed than the old code, which used `len(obj)` to make the decision (bad-sequences with well defined `len`).

fAndreuzzi commented:
Thank you @seberg, unfortunately there is a segmentation fault in `test_conversion_utils`, I tried to reproduce the bug on my computer but it does not show up.

mattip commented:
You uncovered an unsupported feature in PyPy: it does not like format widths. Please skip the crashing test on that platform with a `@pytest.mark.skipif` decorator.

See [PyPy issue 3593](https://foss.heptapod.net/pypy/pypy/-/issues/3593).

fAndreuzzi commented:
@mattip Thank you very much, there is probably more than one test involved (all the tests that pass an unacceptable parameter to `PyArray_IntpConverter` or `PyArray_IntpFromSequence`).

fAndreuzzi commented:
Since skipping all the tests requires changing multiple lines in multiple files in order to add `@pytest.mark.skipif(IS_PYPY, reason="PyPy format broken when using width")` where it is needed, I would suggest to show a different output (i.e. not using the function which is causing problems) when the platform is PyPy, until the error is fixed. What do you think @mattip @seberg? Is this doable in C?

mattip commented:
I think it makes more sense to skip PyPy on CI until they fix it. It is a pretty bad bug, and should not block NumPy's progress. You can comment out this section of the CI config file by putting a '#' in front of each line https://github.com/numpy/numpy/blob/bc087bb2b3104547b097e03e669fb5fa77b16f05/.github/workflows/build_test.yml#L205-L216

fAndreuzzi commented:
Thank you @mattip, I skipped PyPy in two places in ebbb9f3 (GH and Azure Pipeline). The other tests are running properly now.

mattip commented:
Now that all tests are passing, can you show before/after benchmarks? There is a guide to [running the benchmarks](https://numpy.org/devdocs/benchmarking.html), I think you want the `compare` option

fAndreuzzi commented:
Hi @mattip, I'm sorry for the delay but I encountered several issues while running the benchmark (probably related with my compiler). 

This is the result: [numpy_benchmark.txt](https://github.com/numpy/numpy/files/7500726/numpy_benchmark.txt)

Please let me know if the format is OK and if this is what you expected.

mattip commented:
Thanks for the benchmarks. I am adding the text file here to save other reviewers a click. Does it seem reasonable that the first benchmark went up and that `ones` `empty` and `zeros` got so much faster?
```
    [79b381f3]       [9ad477d1]
     <main>           <opt2>
+     5.25±0.08μs      5.92±0.06μs     1.13  bench_scalar.ScalarMath.time_addition_pyint('int32')
+      33.2±0.2μs       36.5±0.6μs     1.10  bench_core.VarComplex.time_var(100)
+         187±1μs          205±2μs     1.09  bench_function_base.Sort.time_sort('merge', 'float64', ('sorted_block', 10))
+      62.7±0.1μs         68.1±2μs     1.09  bench_function_base.Median.time_odd
+     2.97±0.04μs       3.22±0.1μs     1.08  bench_ufunc.ArgParsingReduce.time_add_reduce_arg_parsing((array([0., 1.]), axis=0, dtype=None))
+      41.8±0.2μs         45.2±1μs     1.08  bench_scalar.ScalarMath.time_power_of_two('complex64')
+      12.2±0.3μs       13.1±0.1μs     1.07  bench_lib.Nan.time_nanmin(200, 2.0)
+       422±0.7μs         452±10μs     1.07  bench_io.LoadtxtCSVdtypes.time_loadtxt_dtypes_csv('int64', 100)
+      24.4±0.2μs       26.1±0.1μs     1.07  bench_shape_base.Block2D.time_block2d((16, 16), 'uint8', (2, 2))
+     2.21±0.02μs       2.36±0.1μs     1.07  bench_ufunc.ArgParsingReduce.time_add_reduce_arg_parsing((array([0., 1.]), 0, None, array(0.)))
+     6.12±0.03μs       6.48±0.1μs     1.06  bench_scalar.ScalarMath.time_addition_pyint('float16')
+        44.6±1μs       47.3±0.1μs     1.06  bench_ma.UFunc.time_1d(True, True, 1000)
+     1.43±0.01μs      1.52±0.04μs     1.06  bench_core.CountNonzero.time_count_nonzero(2, 100, <class 'numpy.int64'>)
+      39.8±0.1μs         42.1±1μs     1.06  bench_shape_base.Block2D.time_block2d((256, 256), 'uint8', (2, 2))
+     1.15±0.02ms      1.22±0.01ms     1.06  bench_ufunc_strides.Unary.time_ufunc(<ufunc 'tanh'>, 4, 1, 'd')
+     11.7±0.01μs      12.3±0.06μs     1.05  bench_core.CountNonzero.time_count_nonzero_multi_axis(3, 100, <class 'numpy.int8'>)
-     3.46±0.04μs      3.29±0.02μs     0.95  bench_array_coercion.ArrayCoercionSmall.time_array_subok(range(0, 3))
-      19.4±0.1ms       18.5±0.2ms     0.95  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3, 5, 7])
-      50.5±0.6μs       47.9±0.5μs     0.95  bench_ma.UFunc.time_1d(False, True, 10)
-     1.26±0.02μs      1.20±0.01μs     0.95  bench_ufunc.ArgParsing.time_add_arg_parsing((array(1.), array(2.), subok=True))
-      23.4±0.1μs       22.1±0.3μs     0.94  bench_core.Core.time_diag_l100
-      30.2±0.4μs       28.5±0.1μs     0.94  bench_core.CountNonzero.time_count_nonzero_multi_axis(1, 10000, <class 'numpy.int32'>)
-      73.3±0.5μs       69.1±0.2μs     0.94  bench_random.Choice.time_choice(1000000.0)
-      27.7±0.2μs       26.0±0.1μs     0.94  bench_core.CountNonzero.time_count_nonzero_multi_axis(1, 10000, <class 'numpy.int8'>)
-     9.99±0.02μs      9.38±0.04μs     0.94  bench_reduce.AnyAll.time_all_slow
-      22.9±0.3μs       21.3±0.3μs     0.93  bench_lib.Unique.time_unique(200, 90.0)
-      13.0±0.2μs       12.1±0.3μs     0.93  bench_lib.Nan.time_nanmin(200, 90.0)
-         430±3μs          400±7μs     0.93  bench_core.CountNonzero.time_count_nonzero(3, 1000000, <class 'numpy.int16'>)
-      11.9±0.4μs      11.0±0.04μs     0.93  bench_core.CountNonzero.time_count_nonzero_multi_axis(1, 100, <class 'numpy.int64'>)
-      6.09±0.2μs      5.64±0.02μs     0.93  bench_core.CorrConv.time_convolve(50, 100, 'valid')
-        11.5±2μs      10.6±0.06μs     0.92  bench_itemselection.Take.time_contiguous((1000, 2), 'clip', 'complex256')
-     1.41±0.01μs         1.30±0μs     0.92  bench_array_coercion.ArrayCoercionSmall.time_array_subok([1])
-     2.58±0.08μs      2.39±0.05μs     0.92  bench_ufunc.ArgParsingReduce.time_add_reduce_arg_parsing((array([0., 1.]), axis=0, dtype=None, out=array(0.)))
-         867±4ns          798±4ns     0.92  bench_core.Core.time_arange_100
-      11.0±0.2μs      10.1±0.07μs     0.92  bench_core.CountNonzero.time_count_nonzero_axis(1, 100, <class 'numpy.int16'>)
-      15.0±0.4μs      13.5±0.06μs     0.90  bench_lib.Nan.time_nancumsum(200, 90.0)
-       970±100μs        853±0.9μs     0.88  bench_function_base.Sort.time_sort('heap', 'float64', ('sorted_block', 10))
-      12.9±0.6μs      11.3±0.08μs     0.88  bench_core.CountNonzero.time_count_nonzero_multi_axis(1, 100, <class 'numpy.int32'>)
-         217±1μs        189±0.7μs     0.87  bench_function_base.Sort.time_argsort('merge', 'float64', ('sorted_block', 100))
-      6.81±0.2μs      5.93±0.04μs     0.87  bench_records.Records.time_fromstring_w_dtype
-        783±10μs          645±1μs     0.82  bench_ufunc.UFunc.time_ufunc_types('reciprocal')
-      3.21±0.2μs      2.51±0.03μs     0.78  bench_core.Core.time_ones_100
-         946±7ns         421±10ns     0.45  bench_core.Core.time_zeros_100
-         933±5ns          378±2ns     0.41  bench_core.Core.time_empty_100
```

fAndreuzzi commented:
@mattip Since I do not know the codebase as I only touched a few files I cannot say with 100% certainty that this makes sense. The aim of the project was (initially) to speed up calls in which the shape is scalar.

However @seberg and me refactored a little bit the class `conversion_utils.c`, therefore it is possible that some changes in the performance happened also in the non-scalar case.

seberg commented:
The last 3 are definitely relevant, most/all others are probably just random fluctuations (often these seem to be compiler optimization fluctuations, always wanted to try profile guided optimization to stabilize benchmarks, one of these days my gcc seems to be fixed now ;)).

One thing we could do, is add a benchmark similar to `time_empty_100`, but instead of `100` use `(10, 10)`, which should cover the sequence path as well.

fAndreuzzi commented:
@seberg I rebased the commits against the new changes in master, is there something more I should to in order to prepare the PR for the review?

seberg commented:
I think this is ready and I can jump in with anything that should come up also.  We are a bit in a pre-1.22 rush, and I was hoping to not merge myself since I modified things quite a bit.

The PyPy issue also makes me pause slightly, if PyPy needs catching up, I am not in a rush for 1.22 (it is a significant speed improvement for some cases, but also been slow for 2 decades, so...).

fAndreuzzi commented:
Okay, let me know if I can help somehow. Anyway it looks like a new error came out from the rebase on some platforms:
```
_________________ TestGenericAlias.test_getattr[__deepcopy__] __________________

self = <numpy.typing.tests.test_generic_alias.TestGenericAlias object at 0x7fd586a531c0>
name = '__deepcopy__'

    @pytest.mark.parametrize("name", GETATTR_NAMES)
    def test_getattr(self, name: str) -> None:
        """Test that `getattr` wraps around the underlying type,
        aka ``__origin__``.
    
        """
        value = getattr(NDArray, name)
        value_ref1 = getattr(np.ndarray, name)
    
        if sys.version_info >= (3, 9):
>           value_ref2 = getattr(NDArray_ref, name)
E           AttributeError: 'types.GenericAlias' object has no attribute '__deepcopy__'

name       = '__deepcopy__'
self       = <numpy.typing.tests.test_generic_alias.TestGenericAlias object at 0x7fd586a531c0>
value      = <method '__deepcopy__' of 'numpy.ndarray' objects>
value_ref1 = <method '__deepcopy__' of 'numpy.ndarray' objects>
```

seberg commented:
The test failures just ran into a CI update, re-basing again will fix this.

seberg commented:
Rebased away the pypy related fixups, since I think main is good now.

Planning on squash merging this soon, maybe just later today.  I think it is a definite improvement of both speed and code.  Thanks @fAndreuzzi.

seberg commented:
My bad, the Pypy issue is only fixed in the next release of PyPy (3.7.8), so we probably might as well wait for that.

fAndreuzzi commented:
> Planning on squash merging this soon, maybe just later today. I think it is a definite improvement of both speed and code. Thanks @fAndreuzzi.

Thank you very much @seberg for the kind support and detailed code reviews, I learned a lot from this PR.

> My bad, the Pypy issue is only fixed in the next release of PyPy (3.7.8), so we probably might as well wait for that.

Looking forward to see my code in the main branch!

fAndreuzzi commented:
Hi @seberg, just out of curiosity, how long do you take it's going to take to see the fix to the bug in PyPy published?

seberg commented:
I dunno, before the next release for sure.  In any case, you could also skip the tests as we did in gh-20580.

seberg committed:
Merge branch 'main' into optimize-scalar

seberg commented:
Not sure if our PyPy dependency got bumped, so close/reopen just to see...

seberg committed:
TST: Skip test on older PyPy due to buggy error formatting

seberg committed:
MAINT: Add static to the test functions definition

seberg committed:
TST: More tests marked skip for PyPy (missed one path the first time)

seberg commented:
OK, I marked all the tests that fail on PyPy now.  We could of course just bump the CI PyPy as well, but I guess it also just doesn't hurt to have those skips for now (we already have the similar skips anyway).

Going to finally merge once CI is done.  Thanks again.

fAndreuzzi commented:
Thank you for your help @seberg
